### ISSUE TITLE
Error: Provider produced inconsistent final plan for null_resource after modifying VM

### ISSUE THREAD

**Issue by gdsotirov** [hashicorp/terraform-provider-google#10771]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Problem description

In a configuration with VM instance and null resource with local-exec provisioner triggered by IP change provider fails to generate consistent final plan when IP of VM instance changes after modification. There is no such problem if after modification the public IP of the instance remains the same.

### Terraform Version

```
% terraform --version
Terraform v1.1.1
on darwin_amd64
```

### Affected Resource(s)

* `google_compute_instance` (see [documentation](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_instance));
* `null_resource` (see [documentation](https://registry.terraform.io/providers/hashicorp/null/latest/docs/resources/resource)).

### Terraform Configuration Files

```tf
resource "google_service_account" "default" {
	account_id   = "service_account_id"
	display_name = "Service Account"
}

resource "google_compute_instance" "my-vm" {
	name		= "my-vm"
	machine_type= "n2-standard-2"

	allow_stopping_for_update	= true

	boot_disk {
		initialize_params {
			image	= "centos-cloud/centos-8"
		}
	}

	network_interface {
		network = "default"
	
		access_config {
			// Ephemeral public IP
		}
	}

	service_account {
		email  = google_service_account.default.email
		scopes = ["cloud-platform"]
	}

	# Wait for the instance to become available
	desired_status	= "RUNNING"
}

resource "null_resource" "vm-setup" {
	depends_on	=	[
		google_compute_instance.my-vm
	]

	triggers = {
		my_vm_ipaddr = google_compute_instance.my-vm.network_interface.0.access_config.0.nat_ip
	}

	provisioner "local-exec" {
		command	= "ansible-playbook --inventory '${local.host_ip},' my-vm-setup.yml"
	}
}
```

### Expected Behavior

Provider should produce consistent final plan and there should not be error.

### Actual Behavior

After modifying the VM instance Terraform exits with the following error:

```
Error: Provider produced inconsistent final plan

When expanding the plan for null_resource.my-vm-setup[0] to include new
values learned so far during apply, provider
"registry.terraform.io/hashicorp/null" produced an invalid new value for
.triggers["my_vm_ipaddr"]: was cty.StringVal("1.2.3.4"), but now
cty.StringVal("5.6.7.8").

This is a bug in the provider, which should be reported in the provider's
own issue tracker.
```

### Steps to Reproduce

1. Change a parameter (e.g. `machine_type`) in VM's configuration; and
2. Then run `terraform apply`.

### References

I was not able to find similar issue, so I created this one.

b/308755672

**Comment by ScottSuarez** [comment]
This value is computed from the api.. Unfortunately there is not much we can do here, terraform is not behaving in any unexpected way. Can you switch to not using an Ephemeral IP?

**Comment by gdsotirov** [comment]
@ScottSuarez I'm not sure to understand you (or at least completely). You mean the problem is not solvable?

Yes, I could use Static IPs, but I do not think this is the point of this issue. I'm using Ephemeral IPs for development resources, because Static IPs are really not necessary for these.

**Comment by ScottSuarez** [comment]
You are changing the instance.. terraform reconciles this change and under the hood a value changes that terraform didn't intend to affect. This is the result of the api and how terraform behaves. Terraform is not playing well with this API. It's not an issue fundamental to the provider.

**Comment by gdsotirov** [comment]
Yes, I'm changing the instance, but with Ephemeral IPs the change of instance's IP address is possible and should thus be anticipated. Should I report this to Terraform or should I understand that it is totally unsolvable?

**Comment by ScottSuarez** [comment]
I would suggest opening a ticket against terraform.. They may have advice, or you could open a ticket against the api asking if the field would be able to be consistent across updates.

**Comment by gdsotirov** [comment]
Issue [30696](https://github.com/hashicorp/terraform/issues/30696) opened upstream, so closing this one.

**Comment by gdsotirov** [comment]
Well, upstream responded that it is a "_canonical example of "confusing errors from downstream operations", where the resource unexpectedly changed it's result from the plan, causing errors for other resources_". What do you think @ScottSuarez ?

**Comment by gdsotirov** [comment]
> From this block, if my understanding correct, you are changing the affected resource out of band?

We use Ansible to modify VM instance setup (configurations, installed software, etc.), but this does not include its IP address if this is what you mean by "_out of band_". The provisioning in GCP is done only with Terraform.

> Is there no way to avoid running this script within the apply. Perhaps you could reconcile the resource outside the apply.

Without running Ansible our infrastructure would not be completely prepared to do what it is intended to. Our implementation is based on examples like [Ansible and Terraform - Better Together](https://github.com/scarolan/ansible-terraform) as we want everything provisioned and setup properly from a single starting point. Apart from the trigger by IP address we also have other triggers, so we could start Ansible when necessary to update VM instance setup. Thus it would not be easy to re-implement this outside Terraform apply.

> You might also be able to to utilize a lifecycle rule on the resource in question to ignore changes on this field.

Our point is exactly the opposite - i.e. if VM instance's IP address changed this signals some modification, so we need to re-run Ansible to account for it.

Anyway, we only encounter this issue infrequently and the workaround is to run Terraform again, so if it cannot be fixed easily I'm OK to close it.

**Comment by karolgorc** [comment]
Hi @edwardmedia. I think this issue should be closed. 

After testing this i'd say that using `trigger` on a `nat_ip` field that's dynamically changed shouldn't be supported. This change is happening out of bounds and in this case it goes something like this.

- null provider recognizes a change in the nat_ip but it wasn't fetched from Google API yet so it can't be assigned
- Google provider fetches the changed IP from the API and assigns it to state
- The produced plan is inconsistent
- The second plan already has the new `nat_ip` assigned so it goes through

Basically my point is that `trigger` shouldn't be tied to a value that's changing outside of terraform

**Comment by ScottSuarez** [comment]
ack, this makes sense to me. I am okay to close this issue. Thank you for looking into it.
